### PR TITLE
Use terraform remote state.

### PR DESCRIPTION
**Pull request by ouphi**
- Use terraform remote state instead of publishing state as an artifact.
- Create backend.conf file containing remote state location & publish it as an artifact, to use it for the destroy.
- Delete blob containing the terraform state after destroying terraform resources.

Change on initialize.yaml tested here by changing the trigger temporary: https://github.com/agera-edc/MinimumViableDataspace/runs/6107893856?check_suite_focus=true

### CODE REVIEW

**Review by algattik — APPROVED** [review]

File: .github/workflows/initialize.yaml [hunk 1]
```diff
@@ -16,6 +16,8 @@ jobs:
     env:
       COMMON_RESOURCE_GROUP: ${{ secrets.COMMON_RESOURCE_GROUP }}
       COMMON_RESOURCE_GROUP_LOCATION: ${{ secrets.COMMON_RESOURCE_GROUP_LOCATION }}
+      TERRAFORM_STATE_STORAGE_ACCOUNT: ${{ secrets.TERRAFORM_STATE_STORAGE_ACCOUNT }}
+      TERRAFORM_STATE_CONTAINER: ${{ secrets.TERRAFORM_STATE_CONTAINER }}
```
Comment (cpeeyush) [inline .github/workflows/initialize.yaml:20]: Information about these secrets needs to be updated in [doc](https://github.com/agera-edc/MinimumViableDataspace/blob/main/docs/developer/continuous_deployment.md).

File: .github/workflows/initialize.yaml [hunk 2]
```diff
@@ -29,5 +31,10 @@ jobs:
       - name: "Create resource group"
         run: az group create --name "$COMMON_RESOURCE_GROUP" --location "$COMMON_RESOURCE_GROUP_LOCATION" -o table
 
+      - name: "Create terraform state storage account and container"
+        run: |
+          az storage account create --name ${TERRAFORM_STATE_STORAGE_ACCOUNT} --resource-group ${COMMON_RESOURCE_GROUP}
+          az storage container create --name ${TERRAFORM_STATE_CONTAINER} --account-name ${TERRAFORM_STATE_STORAGE_ACCOUNT}
+
```
Comment (cpeeyush) [inline .github/workflows/initialize.yaml:38]: will this fail if we re-run the pipeline and storage account and container already exists ?

File: .github/workflows/cd.yaml [hunk 3]
```diff
@@ -153,17 +158,25 @@ jobs:
...
+          client-id: ${{ secrets.ARM_CLIENT_ID }}
+          tenant-id: ${{ secrets.ARM_TENANT_ID }}
+          subscription-id: ${{ secrets.ARM_SUBSCRIPTION_ID }}
       - uses: actions/download-artifact@v3
         name: 'Download terraform folder'
         with:
-          name: terraform-${{ matrix.participant }}
-          path: deployment/terraform
+          name: terraform-${{ matrix.participant }}-backend-conf
+          path: deployment/terraform/
```
Comment (cpeeyush) [inline .github/workflows/cd.yaml:173]: nit: maybe we can use the full path to make sure we only get backend.conf downloaded here 
```suggestion
          path: deployment/terraform/backend.conf
```

File: .github/workflows/cd.yaml [hunk 4]
```diff
@@ -153,17 +158,25 @@ jobs:
...
 
       - name: 'Delete terraform resources'
         run: |
-          : # Make terraform provider binaries executable because file permissions of uploaded artifacts are not maintained.
-          chmod -R +x .terraform/providers
-          terraform destroy -var-file=input.tfvars -auto-approve
+          terraform init -backend-config=backend.conf
+          terraform destroy -auto-approve
+          az storage blob delete --account-name ${{ secrets.TERRAFORM_STATE_STORAGE_ACCOUNT }} -c ${{ secrets.TERRAFORM_STATE_CONTAINER }} -n ${{ matrix.participant }}${{ github.run_number }}.tfstate
```
Comment (cpeeyush) [inline .github/workflows/cd.yaml:179]: nit: maybe we can move state file key deletion to its own step for separation. e.g. Delete terraform state

File: .github/workflows/initialize.yaml [hunk 5]
```diff
@@ -29,5 +31,10 @@ jobs:
       - name: "Create resource group"
         run: az group create --name "$COMMON_RESOURCE_GROUP" --location "$COMMON_RESOURCE_GROUP_LOCATION" -o table
 
+      - name: "Create terraform state storage account and container"
```
Comment (cpeeyush) [inline .github/workflows/initialize.yaml:34]: Can you please also add a link to PR description for pipeline which verified that these changes works? As initialize pipeline is not part of PR checks.

Comment (ouphi) [reply .github/workflows/initialize.yaml:34]: It is already in the PR description: https://github.com/agera-edc/MinimumViableDataspace/runs/6107893856?check_suite_focus=true

File: .github/workflows/initialize.yaml (hunk 2 above)
Comment (ouphi) [reply .github/workflows/initialize.yaml:38]: I tested it and it does not fail if the resource already exists

File: .github/workflows/initialize.yaml (hunk 1 above)
Comment (ouphi) [reply .github/workflows/initialize.yaml:20]: Good point thanks! I added them to the doc in `Configure CD settings`.

File: .github/workflows/initialize.yaml (hunk 5 above)
Comment (cpeeyush) [reply .github/workflows/initialize.yaml:34]: Thanks, I missed it.

**Review by cpeeyush — APPROVED** [review]

**Pull request merged**